### DISCUSSION TITLE
Off-line STT ONNX example ❓ Questions / Help / Support

### DISCUSSION THREAD

**Discussion by andy8025** [snakers4/silero-models#182] [Q&A] [answered]
## ❓ Questions and Help

Hi! First, thanks for making this awesome software available. 

I am trying to modify the example shown here  [https://github.com/snakers4/silero-models#onnx](url) to work without omegaconf and torch.hub.load(), just want to load the model directly from a .onnx file.  But I'm stuck at the second to last line - using the decoder function or class. How do I properly use the decoder?

--------------------------------------

import onnx
import torch
import onnxruntime
import utils

language = 'en'

#_, decoder, utils = torch.hub.load(repo_or_dir='snakers4/silero-models', model='silero_stt', language=language)

onnx_model = onnx.load('en_v6_xlarge.onnx')
onnx.checker.check_model(onnx_model)
ort_session = onnxruntime.InferenceSession('en_v6_xlarge.onnx', providers=['CUDAExecutionProvider'])

test_files = ['speech_orig.wav']

batches = utils.split_into_batches(test_files, batch_size=10)
input = utils.prepare_model_input(utils.read_batch(batches[0]))

onnx_input = input.detach().cpu().numpy()
ort_inputs = {'input': onnx_input}
ort_outs = ort_session.run(None, ort_inputs)  

decoded = decoder(torch.Tensor(ort_outs[0])[0])  # need to replace this
decoder = utils.Decoder()  # with ???
print(decoded)`

**Comment by snakers4** [comment, marked as answer]
Please see the provided colab examples and wiki - https://github.com/snakers4/silero-models#onnx

**Comment by zoldaten** [comment]
@andy8025 
colab not working on Windows with torch.hub load problem:
`urllib.error.URLError: <urlopen error [WinError 10060] `

so decoder could be put "by hands":
```
import onnx
import torch
import onnxruntime
from omegaconf import OmegaConf
#https://huggingface.co/Derur/silero-models/tree/main/stt/ua/ua_v3

from utils import (init_jit_model, #utils from silero
                       split_into_batches,
                       read_batch,
                       prepare_model_input, Decoder)
#https://huggingface.co/Derur/silero-models/tree/main
#pip install torchaudio soundfile

# load provided utils
#_, decoder, utils = torch.hub.load(repo_or_dir='snakers4/silero-models', model='silero_stt', language=language, trust_repo=True)
#(read_batch, split_into_batches, read_audio, prepare_model_input) = utils
decoder='ua_v1_labels.json'

 # see available models
#torch.hub.download_url_to_file('https://raw.githubusercontent.com/snakers4/silero-models/master/models.yml', 'models.yml')
models = OmegaConf.load('models.yml')
available_languages = list(models.stt_models.keys())
print(available_languages)
#assert language in available_languages

# load the actual ONNX model
#torch.hub.download_url_to_file(models.stt_models.en.latest.onnx, 'model.onnx', progress=True)
onnx_model = onnx.load('ua_v3.onnx')
onnx.checker.check_model(onnx_model)
ort_session = onnxruntime.InferenceSession('ua_v3.onnx')


# download a single file, any format compatible with TorchAudio
#torch.hub.download_url_to_file('https://opus-codec.org/static/examples/samples/speech_orig.wav', dst ='speech_orig.wav', progress=True)
test_files = ['voice.wav']
batches = split_into_batches(test_files, batch_size=10)
input = prepare_model_input(read_batch(batches[0]))

# actual onnx inference and decoding
onnx_input = input.detach().cpu().numpy()
ort_inputs = {'input': onnx_input}
ort_outs = ort_session.run(None, ort_inputs)
#print(torch.Tensor(ort_outs[0])[0])
torch.set_grad_enabled(False)
model = torch.jit.load('ua_v3_jit.model', map_location='cpu')
model.eval()
#_, decoder = init_jit_model(model_url=models.stt_models.get(language).get(version).get('jit'),)
#https://github.com/snakers4/silero-models/blob/master/src/silero/utils.py#L50

decoder=Decoder(model.labels)
decoded = decoder(torch.Tensor(ort_outs[0])[0])
print(decoded)
```
